sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java

```diff
@@ -0,0 +1,211 @@
+import org.jruby.javasupport.JavaEmbedUtils;
+import org.jruby.runtime.ThreadContext;
+import org.jruby.runtime.builtin.IRubyObject;
+import org.sonarsource.ruby.converter.adapter.NodeAdapter;
+import org.sonarsource.ruby.converter.adapter.RangeAdapter;
+import org.sonarsource.ruby.converter.adapter.SourceMapAdapter;
+import org.sonarsource.slang.api.IdentifierTree;
+import org.sonarsource.slang.api.Tree;
+import org.sonarsource.slang.api.TreeMetaData;
+import org.sonarsource.slang.impl.ClassDeclarationTreeImpl;
+import org.sonarsource.slang.impl.IdentifierTreeImpl;
+import org.sonarsource.slang.impl.LiteralTreeImpl;
+import org.sonarsource.slang.impl.NativeTreeImpl;
+import org.sonarsource.slang.impl.TreeMetaDataProvider;
+
+@JRubyClass(name = "RubyVisitor")
+public class RubyVisitor extends RubyObject {
+
+  private static final String AST_PROCESSOR_CLASS = "Parser::AST::Processor";
+  private static RubyClass superclass = null;
+
+  private transient TreeMetaDataProvider metaDataProvider;
+
+  public RubyVisitor(Ruby runtime, RubyClass metaClass) {
+    super(runtime, metaClass);
+  }
+
+  public static void addToRuntime(final Ruby runtime) {
+    superclass = runtime.getObject().subclasses(true).stream()
+      .filter(clazz -> AST_PROCESSOR_CLASS.equals(clazz.getName()))
+      .findFirst()
+      .orElseThrow(() -> new IllegalStateException(String.format("Could not find class %s in ruby runtime", AST_PROCESSOR_CLASS)));
+    RubyModule rubyProcessor = runtime.getObject().defineClassUnder(RubyVisitor.class.getSimpleName(), superclass, RubyVisitor::new);
+    rubyProcessor.defineAnnotatedMethods(RubyVisitor.class);
+  }
+
+  /**
+   * Ruby constructor takes 2 parameters, metaDataProvider as first parameter, and superclass as second
+   */
+  @JRubyMethod(name = "new", required = 1, rest = true, meta = true)
+  public static IRubyObject rbNew(ThreadContext context, IRubyObject klazz, IRubyObject[] args) {
+    RubyVisitor rubyVisitor = (RubyVisitor) ((RubyClass) klazz).allocate();
+    rubyVisitor.metaDataProvider = args[0].toJava(TreeMetaDataProvider.class);
+    return rubyVisitor;
+  }
+
+  @JRubyMethod
+  public IRubyObject process(ThreadContext context, IRubyObject node) {
+    // some nodes present in the AST but are empty (e.g. function parameters clause)
+    // or node is nil (absent)
+    if (noLocation(node) || node.isNil()) {
+      return new RubyNil(getRuntime());
+    }
+
+    IRubyObject processedNode = callSuperMethod(context, "process", node);
+
+    if (processedNode instanceof NodeAdapter) {
+      // A more specific tree was already created by a specialized method (on_${type}, ex: on_int)
+      return processedNode;
+
+    } else {
+      return createNativeTreeAdapter(processedNode);
+    }
+  }
+
+  @JRubyMethod(name = "on_class")
+  public IRubyObject onClass(ThreadContext context, IRubyObject node) {
+    IRubyObject updatedNode = callSuperMethod(context, "on_class", node);
```

Comment:
maybe we can rename `callSuperMethod`  to `continueVisit` , so it is more clear what it's doing

Reply:
Renamed method to `visitNode`